src/adapters/thinkerbell/mod.rs

```diff
@@ -0,0 +1,404 @@
+
+    fn send_values(&self, values: HashMap<Id<Setter>, Value>) -> ResultMap<Id<Setter>, (), Error> {
+        values.iter()
+            .map(|(id, value)| {
+                let (tx, rx) = channel();
+                let _ = self.tx.send(ThinkAction::RespondToSetter(tx, id.clone(), value.clone()));
+                match rx.recv() {
+                    Ok(result) => (id.clone(), result),
+                    // If an error occurs, the channel died!
+                    Err(recv_err) => (id.clone(), Err(Error::InternalError(
+                            InternalError::GenericError(format!("{:?}", recv_err))))),
+                }
+            })
+            .collect()
+    }
+
+    fn register_watch(&self, mut watch: Vec<(Id<Getter>, Option<Range>)>,
+        _: Box<ExtSender<WatchEvent>>) ->
+            ResultMap<Id<Getter>, Box<AdapterWatchGuard>, Error> {
+        watch.drain(..).map(|(id, _)| {
+            (id.clone(), Err(Error::GetterDoesNotSupportWatching(id)))
+        }).collect()
+    }
+}
+
+/// ThinkerbellAdapter's main loop handles messages of these types.
+enum ThinkAction {
+    AddRuleService(Id<ScriptId>),
+    RemoveRuleService(Id<ScriptId>),
+    RespondToGetter(RawSender<Result<Option<Value>, Error>>, Id<Getter>),
+    RespondToSetter(RawSender<Result<(), Error>>, Id<Setter>, Value),
+}
+
+/// An internal data structure to track getters and setters.
+struct ThinkerbellRule {
+    script_id: Id<ScriptId>,
+    service_id: Id<ServiceId>,
+    getter_source_id: Id<Getter>,
+    getter_is_enabled_id: Id<Getter>,
+    setter_is_enabled_id: Id<Setter>,
+    setter_remove_id: Id<Setter>,
+}
+
+impl<T> ThinkerbellAdapter<T> where T: AdapterManagerHandle + API + Clone + 'static {
+
+    fn main(
+        &self,
+        rx: Receiver<ThinkAction>,
+        mut script_manager: ScriptManager<ThinkerbellExecutionEnv<T>, RawSender<(Id<ScriptId>, ExecutionEvent)>>
+    ) {
+        // Store an in-memory list of all of the rules (their getters, setters, etc.).
+        // We need to track these to respond to getter/setter requests.
+        let mut rules = Vec::new();
+
+        'recv: while let Ok(action) = rx.recv() {
+            match action {
+                // After a script has been started (elsewhere), start a Service for that script.
+                ThinkAction::AddRuleService(script_id) => {
+                    match self.add_rule_service(script_id) {
+                        Ok(rule) => {
+                            rules.push(rule);
+                        },
+                        Err(e) => {
+                            error!("Unable to add Thinkerbell Rule Service: {:?}", e);
+                        }
+                    };
+                },
+                // After a script has been stopped (elsewhere), remove the Service for that script.
+                ThinkAction::RemoveRuleService(script_id) => {
+                    for ref rule in &rules {
+                        if rule.script_id == script_id {
```

Comment:
These lookups in the full vector don't feel me with hope for the future. I'd rather have a `HashMap` lookup.

This can wait for a followup, if you do it soon.